src/gossip/service.zig

```diff
@@ -497,62 +518,26 @@ pub const GossipService = struct {
         var pull_responses = try ArrayList(PullResponseMessage).initCapacity(self.allocator, init_capacity);
         defer pull_responses.deinit();
 
-        var prune_messages = try ArrayList(*PruneData).initCapacity(self.allocator, init_capacity);
+        var prune_messages = try ArrayList(*const PruneData).initCapacity(self.allocator, init_capacity);
         defer prune_messages.deinit();
 
-        while (!self.exit.load(.unordered)) {
-            const maybe_messages = try self.verified_incoming_channel.try_drain();
-
-            if (maybe_messages == null) {
-                continue;
-            }
-
-            const messages = maybe_messages.?;
-            defer {
-                for (messages) |*msg| {
-                    // Important: this uses shallowFree instead of bincode.free
-                    //
-                    // The message contains some messaging metadata plus a
-                    // payload of a SignedGossipData. The metadata won't be needed
-                    // after this iteration is complete. The payload will be
-                    // needed since it is stored in the GossipTable.
-                    //
-                    // bincode.free would free the entire message including the
-                    // payload. This would lead to a segfault if the data is
-                    // accessed from the GossipTable later.
-                    //
-                    // Not freeing at all would lead to a memory leak of any
-                    // allocations in the metadata.
-                    //
-                    // The compromise is a "shallow" free that only frees the
-                    // messaging metadata. SignedGossipData ownership will be
-                    // transferred to GossipTable. The GossipTable implementation
-                    // becomes responsible for freeing any SignedGossipDatas when
-                    // needed.
-                    //
-                    // TODO: this approach is not ideal because it is difficult
-                    // to maintain. Another approach such as reference counting
-                    // would be safer. For more info, see:
-                    // - GossipTable.remove
-                    // - https://github.com/Syndica/sig/pull/69
-                    msg.message.shallowFree(self.gossip_value_allocator);
-                }
-                self.verified_incoming_channel.allocator.free(messages);
-            }
-
-            msg_count += messages.len;
-
-            for (messages) |*message| {
+        // keep waiting for new data until,
+        // - `exit` isn't set,
+        // - there isn't any data to process in the input channel, in order to block the join until we've finished
+        while (!(self.exit.load(.acquire) and self.verified_incoming_channel.len() == 0)) {
+            var msg_count: usize = 0;
+            while (self.verified_incoming_channel.receive()) |message| {
+                msg_count += 1;
```

Comment:
can we have a max msg_count in this loop so it wont get get stuck draining the channel

Reply:
Fair point!